### ISSUE TITLE
Fully silent update in C# (Xamarin Forms on Mac)

### ISSUE THREAD

**Issue by ech0matrix** [sparkle-project/Sparkle#1209]
I've been trying to get silent updates working on my Xamarin Forms app. I'm using [Sparkle-Sharp](https://github.com/rainycape/SparkleSharp) for my C# bindings to Sparkle, and using the information in issues [1047](https://github.com/sparkle-project/Sparkle/issues/1047) and [787](https://github.com/sparkle-project/Sparkle/issues/787) to get me to where I am now.

The CheckForUpdatesInBackground doesn't seem to be working for me, so I've been keeping my own timer and occasionally calling InstallUpdatesIfAvailable. This has worked well to download and install updates without any user intervention (my requirement). The progress bar on the download is not desirable, but can be overlooked. However, every time my loop invokes InstallUpdatesIfAvailable, there's a popup showing the progress of the update check (even if there's no new updates). This isn't a good user experience, and I'm trying to figure out how to not show any popups (completely silent and automatic, without user intervention.)

My setup looks like this in C#:
`
NSUserDefaults.StandardUserDefaults.SetBool(true, "SUAutomaticallyUpdate");
NSUserDefaults.StandardUserDefaults.Synchronize();
SUUpdater.SharedUpdater.AutomaticallyDownloadsUpdates = true;
SUUpdater.SharedUpdater.AutomaticallyChecksForUpdates = true;
`

And this in my info.plist:
`
SUAutomaticallyUpdate Boolean YES
SUEnableAutomaticChecks Boolean YES
SUScheduleCheckInterval Number 3600
SUPublicDSAKeyFile String xxx.pem
SUFeedURL String https://xxx/macosclient.xml
SUShowReleaseNotes Boolean NO
`

Then I have a background worker that makes this call every hour:
`
Device.BeginInvokeOnMainThread(SUUpdater.SharedUpdater.InstallUpdatesIfAvailable);
`

I've tried a few combinations of settings, but this gets me as close to the behavior as I'm looking for. Any ideas on how to hide the popup that occurs while Sparkle is checking for updates? Bonus points for also hiding the download popup to make this truly silent.

**Comment by kornelski** [comment]
Sounds like the same problem as Qt. We depend on Cocoa to listen for app termination to perform install. If you quit the app any other way we don't install anything.

**Comment by ech0matrix** [comment]
Thanks for the quick response! I tried a few things out just now, and it looks like if I use CheckForUpdatesInBackground instead, the update is downloaded silently. When I restart, it's the new version, and I've seen no popups. So that's good, except I need to be able to force the update immediately after it downloads (restarting the app) instead of waiting for the user to close the app. Is there a way to do this.

**Comment by kornelski** [comment]
There's a delegate method that gives you a callback to force restart.

**Comment by ech0matrix** [comment]
@kornelski, a delegate that let's you know when an update has downloaded and is ready for the app to restart? Which one is that? willInstallUpdate?

**Comment by kornelski** [comment]
Yes, `SUUpdaterDelegate::updater:willInstallUpdateOnQuit:immediateInstallationInvocation` and invoke the callback it gives.

**Comment by ech0matrix** [comment]
It sounds like this will initiate the restart. However, how do I know when to call it? How do I know when the update is ready for the restart?

**Comment by ech0matrix** [comment]
I understand that a restart is required. Is there a boolean that I can check to see if Sparkle is waiting for a restart?

**Comment by ech0matrix** [comment]
Oh, okay. I misunderstood. Is there a method then that I can invoke to cause my application to restart?

**Comment by kornelski** [comment]
Is invocation of the callback not causing a restart?

**Comment by gaocan1992** [comment]
Use this delegate callback

```
public override void UpdaterWillInstallUpdateOnQuit(SUUpdater updater, SUAppcastItem item, NSInvocation invocation)
 {
     invocation.Invoke();
 }
```

**Comment by gaocan1992** [comment]
@ech0matrix Can I ask you if you are running into this problem https://github.com/sparkle-project/Sparkle/issues/1203

**Comment by ech0matrix** [comment]
@gaocan1992 I don't think so. I'm verifying automaticallyDownloadsUpdates is true before running the background check.

**Comment by gaocan1992** [comment]
@ech0matrix May I ask how do you verify this value and what if this value is not true.

**Comment by gaocan1992** [comment]
```
- (BOOL)automaticallyDownloadsUpdates
{
    // If we aren't allowed automatic updates, don't ever let them happen
    if (![SUSystemUpdateInfo systemAllowsAutomaticUpdatesForHost:self.host]) {
        return NO;
    }

    // Otherwise, automatically downloading updates is allowed. Does the user want it?
    return [self.host boolForUserDefaultsKey:SUAutomaticallyUpdateKey];
}
```

**Comment by ech0matrix** [comment]
@gaocan1992 
My use case is that we want the application to check for updates, download, install, and restart, with 0 user interaction. We're using C# (Xamarin Forms on Mac). I'm using Sparkle-Sharp for C# bindings. The updater object doesn't seem to do anything unless I invoke it manually (since I didn't make an updater object in Objective C).

Right now, I have a background worker that periodically invokes the updater to check for updates. It gets initialized partly in C#, and partly in the info.plist. You can see my setup and code in my original post at the top of this thread. I'm forcing "AutomaticallyDownloadsUpdates = true;" before my worker calls InstallUpdatesIfAvailable.

Unfortunately, InstallUpdatesIfAvailable still generates popups for progress bars on the check and install, so the whole point of this thread was to try to find a quiet way. It sounds like my solution will be to use CheckForUpdatesInBackground, and setup a delegate to trigger the restart. I'm hoping this will solve my problem, but I won't be able to try it out until next week.

**Comment by gaocan1992** [comment]
@ech0matrix Seems like I have the same setup. @kornelski Suggest me don't force AutomaticallyDownloadsUpdates = true instead setting it in Info.plist. 

And I have set this values in my Info.plist already
```
        <key>SUEnableAutomaticChecks</key>
	<true/>
	<key>SUAllowsAutomaticUpdates</key>
	<true/>
	<key>SUAutomaticallyUpdate</key>
	<true/>
```

I use CheckForUpdatesInBackground and use the delegate callback to trigger the restart.
However, I still see the popup sometimes which is weird. 

By reading the source code I noticed that before Sparkle chooses SUAutomaticUpdateDriver which is the UI less update driver, it will check "automaticallyDownloadsUpdates" which is read from SUHost (return [self.host boolForUserDefaultsKey:SUAutomaticallyUpdateKey];) and SUHost will read the settings from Info.plist.

So I think even if you force AutomaticallyDownloadsUpdates = true it will try to modify the Info.plist
for the value of SUAutomaticallyUpdateKey but when CheckForUpdatesInBackground tries to decide the value of this AutomaticallyDownloadsUpdates it still have to read from the value of SUAutomaticallyUpdateKey.

Could you mention me if you also see UI even you use CheckForUpdatesInBackground.
I doubt there is a threading issue in Sparkle.

**Comment by ech0matrix** [comment]
@gaocan1992 I do have a call in my C# to set SUAutomaticallyUpdate as well, although I was just trying everything I could to get it to work:
` NSUserDefaults.StandardUserDefaults.SetBool(true, "SUAutomaticallyUpdate");
` NSUserDefaults.StandardUserDefaults.Synchronize();

As I mentioned, I'll be able to play around more with the CheckForUpdatesInBackground next week. But so far, I have not seen the UI when using that function. In fact, I did not even know it was doing anything at all, but I manually restarted the app and noticed the version number had bumped.

**Comment by gaocan1992** [comment]
@ech0matrix This should be the right behavior.  I may look into this to make sure that the bool value is true before actually call CheckForUpdatesInBackground, thanks
`NSUserDefaults.StandardUserDefaults.SetBool(true, "SUAutomaticallyUpdate"); NSUserDefaults.StandardUserDefaults.Synchronize();`

**Comment by gaocan1992** [comment]
@ech0matrix Everything looks good, but why do you need put CheckForUpdatesInBackground in a loop?
I believe you can set CheckInterval to 1 hour.

**Comment by ech0matrix** [comment]
@gaocan1992 I do have SUScheduledCheckInterval set to 1 hour. I tried letting Sparkle check for updates on its own, but after waiting for a while, I didn't see anything happening. I even tried setting it to something low, like 30 seconds. So I'm manually invoking the check in order to poll for updates. Maybe I'm missing something in my setup steps, but that's how I've gotten it to work.

**Comment by ech0matrix** [comment]
For posterity, I'll mention that in my solution mentioned above, my C# delegate was getting garbage collected, causing null exceptions in release mode. I needed to hang on to a reference of my SparkleDelegate in C# land so that it doesn't disappear.

`private SparkleDelegate _updaterDelegates = new SparkleDelegate();`

`SUUpdater.SharedUpdater.Delegate = _updaterDelegates;`
